src/libsyntax/test.rs

```diff
@@ -343,288 +325,65 @@ enum BadTestSignature {
-}
-
-/*
-
-We're going to be building a module that looks more or less like:
-
-mod __test {
-  extern crate test (name = "test", vers = "...");
-  fn main() {
-    test::test_main_static(&::os::args()[], tests, test::Options::new())
-  }
-
-  static tests : &'static [test::TestDescAndFn] = &[
-    ... the list of tests in the crate ...
-  ];
-}
-
-*/
-
-fn mk_std(cx: &TestCtxt) -> P<ast::Item> {
-    let id_test = Ident::from_str("test");
-    let sp = ignored_span(cx, DUMMY_SP);
-    let (vi, vis, ident) = if cx.is_libtest {
-        (ast::ItemKind::Use(P(ast::UseTree {
-            span: DUMMY_SP,
-            prefix: path_node(vec![id_test]),
-            kind: ast::UseTreeKind::Simple(None, ast::DUMMY_NODE_ID, ast::DUMMY_NODE_ID),
-        })),
-         ast::VisibilityKind::Public, keywords::Invalid.ident())
-    } else {
-        (ast::ItemKind::ExternCrate(None), ast::VisibilityKind::Inherited, id_test)
-    };
-    P(ast::Item {
-        id: ast::DUMMY_NODE_ID,
-        ident,
-        node: vi,
-        attrs: vec![],
-        vis: dummy_spanned(vis),
-        span: sp,
-        tokens: None,
-    })
-}
-
+/// Creates a function item for use as the main function of a test build.
+/// This function will call the `test_runner` as specified by the crate attribute
 fn mk_main(cx: &mut TestCtxt) -> P<ast::Item> {
     // Writing this out by hand with 'ignored_span':
     //        pub fn main() {
     //            #![main]
-    //            use std::slice::AsSlice;
-    //            test::test_main_static(::std::os::args().as_slice(), TESTS, test::Options::new());
+    //            test::test_main_static(::std::os::args().as_slice(), &[..tests]);
     //        }
-
     let sp = ignored_span(cx, DUMMY_SP);
     let ecx = &cx.ext_cx;
-
-    // test::test_main_static
-    let test_main_path =
-        ecx.path(sp, vec![Ident::from_str("test"), Ident::from_str("test_main_static")]);
+    let test_id = ecx.ident_of("test").gensym();
 
     // test::test_main_static(...)
-    let test_main_path_expr = ecx.expr_path(test_main_path);
-    let tests_ident_expr = ecx.expr_ident(sp, Ident::from_str("TESTS"));
+    let mut test_runner = cx.test_runner.clone().unwrap_or(
+        ecx.path(sp, vec![
+            test_id, ecx.ident_of("test_main_static")
+        ]));
+
+    test_runner.span = sp;
+
+    let test_main_path_expr = ecx.expr_path(test_runner.clone());
     let call_test_main = ecx.expr_call(sp, test_main_path_expr,
-                                       vec![tests_ident_expr]);
+                                       vec![mk_tests_slice(cx)]);
     let call_test_main = ecx.stmt_expr(call_test_main);
+
     // #![main]
     let main_meta = ecx.meta_word(sp, Symbol::intern("main"));
     let main_attr = ecx.attribute(sp, main_meta);
+
+    // extern crate test as test_gensym
+    let test_extern_stmt = ecx.stmt_item(sp, ecx.item(sp,
+        test_id,
+        vec![],
+        ast::ItemKind::ExternCrate(Some(Symbol::intern("test")))
```

Comment:
If this isn't actually used, could this `extern crate` statement be elided?

Reply:
Ah yeah that should only be generated if the test_runner is the default

Reply:
resolved